### ISSUE TITLE
pip install

### ISSUE THREAD

**Issue by MarkGotham** [adamtuft/c-kernel#1]
Thanks for this @adamtuft !

Attempting pip install raises an error:
`#include <mqueue.h>`
and

```
note: This error originates from a subprocess, and is likely not a problem with pip. 
error: legacy-install-failure
```

Thanks for any tips

**Comment by adamtuft** [comment]
Hi Mark, thanks for posting the issue! Can you share the output of:

```
pip install ipcqueue
```

**Comment by MarkGotham** [comment]
Thanks. Same error code / message.

**Comment by adamtuft** [comment]
Would you share the full output please? Might be something useful in there. Can you also post your python & pip versions please.

**Comment by mbukaea** [comment]
I get the same error. For me it is due to mqueue.h not being available on MacOS https://stackoverflow.com/questions/10079403/cannot-find-include-file-mqueue-h-on-os-x
https://discussions.apple.com/thread/8418262 (more recent comment about it) 
Also If I use 'locate mqueue.h' on my system it doesn't find anything

My pip version is 
pip --version
pip 23.2.1 from /opt/local/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/pip (python 3.10)

My python version is
python --version
Python 3.10.13

The output I get is
bash install-ckernel.sh
Collecting ckernel
  Obtaining dependency information for ckernel from https://files.pythonhosted.org/packages/48/ef/88ffcea9c2950f1197334a5441e62b6a3913e031a526d29adc636e1c243a/ckernel-0.4.1-py3-none-any.whl.metadata
  Downloading ckernel-0.4.1-py3-none-any.whl.metadata (6.2 kB)
Collecting ipcqueue (from ckernel)
  Downloading ipcqueue-0.9.7.tar.gz (16 kB)
  Preparing metadata (setup.py) ... done
Requirement already satisfied: ipykernel in /Users/wf5617/Library/Python/3.10/lib/python/site-packages (from ckernel) (6.25.2)
Requirement already satisfied: jupyter-client in /Users/wf5617/Library/Python/3.10/lib/python/site-packages (from ckernel) (8.3.1)
Collecting colorama (from ckernel)
  Downloading colorama-0.4.6-py2.py3-none-any.whl (25 kB)
Requirement already satisfied: cffi>=1 in /Users/wf5617/Library/Python/3.10/lib/python/site-packages (from ipcqueue->ckernel) (1.15.1)
Requirement already satisfied: appnope in /Users/wf5617/Library/Python/3.10/lib/python/site-packages (from ipykernel->ckernel) (0.1.3)
Requirement already satisfied: comm>=0.1.1 in /Users/wf5617/Library/Python/3.10/lib/python/site-packages (from ipykernel->ckernel) (0.1.4)
Requirement already satisfied: debugpy>=1.6.5 in /Users/wf5617/Library/Python/3.10/lib/python/site-packages (from ipykernel->ckernel) (1.7.0)
Requirement already satisfied: ipython>=7.23.1 in /Users/wf5617/Library/Python/3.10/lib/python/site-packages (from ipykernel->ckernel) (8.15.0)
Requirement already satisfied: jupyter-core!=5.0.*,>=4.12 in /Users/wf5617/Library/Python/3.10/lib/python/site-packages (from ipykernel->ckernel) (5.3.1)
Requirement already satisfied: matplotlib-inline>=0.1 in /Users/wf5617/Library/Python/3.10/lib/python/site-packages (from ipykernel->ckernel) (0.1.6)
Requirement already satisfied: nest-asyncio in /Users/wf5617/Library/Python/3.10/lib/python/site-packages (from ipykernel->ckernel) (1.5.7)
Requirement already satisfied: packaging in /Users/wf5617/Library/Python/3.10/lib/python/site-packages (from ipykernel->ckernel) (22.0)
Requirement already satisfied: psutil in /Users/wf5617/Library/Python/3.10/lib/python/site-packages (from ipykernel->ckernel) (5.9.5)
Requirement already satisfied: pyzmq>=20 in /Users/wf5617/Library/Python/3.10/lib/python/site-packages (from ipykernel->ckernel) (25.1.1)
Requirement already satisfied: tornado>=6.1 in /Users/wf5617/Library/Python/3.10/lib/python/site-packages (from ipykernel->ckernel) (6.3.3)
Requirement already satisfied: traitlets>=5.4.0 in /Users/wf5617/Library/Python/3.10/lib/python/site-packages (from ipykernel->ckernel) (5.9.0)
Requirement already satisfied: python-dateutil>=2.8.2 in /Users/wf5617/Library/Python/3.10/lib/python/site-packages (from jupyter-client->ckernel) (2.8.2)
Requirement already satisfied: pycparser in /Users/wf5617/Library/Python/3.10/lib/python/site-packages (from cffi>=1->ipcqueue->ckernel) (2.21)
Requirement already satisfied: backcall in /Users/wf5617/Library/Python/3.10/lib/python/site-packages (from ipython>=7.23.1->ipykernel->ckernel) (0.2.0)
Requirement already satisfied: decorator in /Users/wf5617/Library/Python/3.10/lib/python/site-packages (from ipython>=7.23.1->ipykernel->ckernel) (5.1.1)
Requirement already satisfied: jedi>=0.16 in /Users/wf5617/Library/Python/3.10/lib/python/site-packages (from ipython>=7.23.1->ipykernel->ckernel) (0.19.0)
Requirement already satisfied: pickleshare in /Users/wf5617/Library/Python/3.10/lib/python/site-packages (from ipython>=7.23.1->ipykernel->ckernel) (0.7.5)
Requirement already satisfied: prompt-toolkit!=3.0.37,<3.1.0,>=3.0.30 in /Users/wf5617/Library/Python/3.10/lib/python/site-packages (from ipython>=7.23.1->ipykernel->ckernel) (3.0.39)
Requirement already satisfied: pygments>=2.4.0 in /Users/wf5617/Library/Python/3.10/lib/python/site-packages (from ipython>=7.23.1->ipykernel->ckernel) (2.16.1)
Requirement already satisfied: stack-data in /Users/wf5617/Library/Python/3.10/lib/python/site-packages (from ipython>=7.23.1->ipykernel->ckernel) (0.6.2)
Requirement already satisfied: exceptiongroup in /Users/wf5617/Library/Python/3.10/lib/python/site-packages (from ipython>=7.23.1->ipykernel->ckernel) (1.1.3)
Requirement already satisfied: pexpect>4.3 in /Users/wf5617/Library/Python/3.10/lib/python/site-packages (from ipython>=7.23.1->ipykernel->ckernel) (4.8.0)
Requirement already satisfied: platformdirs>=2.5 in /Users/wf5617/Library/Python/3.10/lib/python/site-packages (from jupyter-core!=5.0.*,>=4.12->ipykernel->ckernel) (3.10.0)
Requirement already satisfied: six>=1.5 in /Users/wf5617/Library/Python/3.10/lib/python/site-packages (from python-dateutil>=2.8.2->jupyter-client->ckernel) (1.16.0)
Requirement already satisfied: parso<0.9.0,>=0.8.3 in /Users/wf5617/Library/Python/3.10/lib/python/site-packages (from jedi>=0.16->ipython>=7.23.1->ipykernel->ckernel) (0.8.3)
Requirement already satisfied: ptyprocess>=0.5 in /Users/wf5617/Library/Python/3.10/lib/python/site-packages (from pexpect>4.3->ipython>=7.23.1->ipykernel->ckernel) (0.7.0)
Requirement already satisfied: wcwidth in /Users/wf5617/Library/Python/3.10/lib/python/site-packages (from prompt-toolkit!=3.0.37,<3.1.0,>=3.0.30->ipython>=7.23.1->ipykernel->ckernel) (0.2.6)
Requirement already satisfied: executing>=1.2.0 in /Users/wf5617/Library/Python/3.10/lib/python/site-packages (from stack-data->ipython>=7.23.1->ipykernel->ckernel) (1.2.0)
Requirement already satisfied: asttokens>=2.1.0 in /Users/wf5617/Library/Python/3.10/lib/python/site-packages (from stack-data->ipython>=7.23.1->ipykernel->ckernel) (2.4.0)
Requirement already satisfied: pure-eval in /Users/wf5617/Library/Python/3.10/lib/python/site-packages (from stack-data->ipython>=7.23.1->ipykernel->ckernel) (0.2.2)
Downloading ckernel-0.4.1-py3-none-any.whl (20 kB)
Building wheels for collected packages: ipcqueue
  Building wheel for ipcqueue (setup.py) ... error
  error: subprocess-exited-with-error
  
  × python setup.py bdist_wheel did not run successfully.
  │ exit code: 1
  ╰─> [35 lines of output]
      /opt/local/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/setuptools/__init__.py:84: _DeprecatedInstaller: setuptools.installer and fetch_build_eggs are deprecated.
      !!
      
              ********************************************************************************
              Requirements should be satisfied by a PEP 517 installer.
              If you are using pip, you can try `pip install --use-pep517`.
              ********************************************************************************
      
      !!
        dist.fetch_build_eggs(dist.setup_requires)
      running bdist_wheel
      running build
      running build_py
      creating build
      creating build/lib.macosx-13.0-arm64-cpython-310
      creating build/lib.macosx-13.0-arm64-cpython-310/ipcqueue
      copying ipcqueue/sysvmq.py -> build/lib.macosx-13.0-arm64-cpython-310/ipcqueue
      copying ipcqueue/serializers.py -> build/lib.macosx-13.0-arm64-cpython-310/ipcqueue
      copying ipcqueue/__init__.py -> build/lib.macosx-13.0-arm64-cpython-310/ipcqueue
      copying ipcqueue/posixmq.py -> build/lib.macosx-13.0-arm64-cpython-310/ipcqueue
      running build_ext
      generating cffi module 'build/temp.macosx-13.0-arm64-cpython-310/ipcqueue._sysvmq.c'
      creating build/temp.macosx-13.0-arm64-cpython-310
      generating cffi module 'build/temp.macosx-13.0-arm64-cpython-310/ipcqueue._posixmq.c'
      building 'ipcqueue._posixmq' extension
      creating build/temp.macosx-13.0-arm64-cpython-310/build
      creating build/temp.macosx-13.0-arm64-cpython-310/build/temp.macosx-13.0-arm64-cpython-310
      creating build/temp.macosx-13.0-arm64-cpython-310/ipcqueue
      /usr/bin/clang -Wno-unused-result -Wsign-compare -Wunreachable-code -fno-common -dynamic -DNDEBUG -g -fwrapv -O3 -Wall -pipe -Os -isysroot/Library/Developer/CommandLineTools/SDKs/MacOSX13.sdk -I/opt/local/Library/Frameworks/Python.framework/Versions/3.10/include/python3.10 -c build/temp.macosx-13.0-arm64-cpython-310/ipcqueue._posixmq.c -o build/temp.macosx-13.0-arm64-cpython-310/build/temp.macosx-13.0-arm64-cpython-310/ipcqueue._posixmq.o -I/private/var/folders/qb/6pjhrfwn2xz04_zv28d2ydb00000gq/T/pip-install-s5zalcx8/ipcqueue_9cd01180c687486dba1342dc765cb9af
      In file included from build/temp.macosx-13.0-arm64-cpython-310/ipcqueue._posixmq.c:569:
      /private/var/folders/qb/6pjhrfwn2xz04_zv28d2ydb00000gq/T/pip-install-s5zalcx8/ipcqueue_9cd01180c687486dba1342dc765cb9af/ipcqueue/posixmq.h:5:10: fatal error: 'mqueue.h' file not found
      #include <mqueue.h>
               ^~~~~~~~~~
      1 error generated.
      error: command '/usr/bin/clang' failed with exit code 1
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for ipcqueue
  Running setup.py clean for ipcqueue
Failed to build ipcqueue
ERROR: Could not build wheels for ipcqueue, which is required to install pyproject.toml-based projects
/opt/local/bin/python: No module named ckernel
/opt/local/bin/python: No module named ckernel
Available kernels:
  python3    /Users/wf5617/Library/Python/3.10/share/jupyter/kernels/python3

**Comment by adamtuft** [comment]
Thanks very much :+1: it looks like the `ipcqueue` package has a hard requirement for POSIX message queues to be available on a system, which they aren't on macOS. I'll have to switch to a different package which is more widely supported.

Thanks for bringing this to my attention! I'll fix this in a future release. I'll leave this issue open until then.

**Comment by Da-byte** [comment]
Adding here that I am having the same issue on Windows 11.

note: This error originates from a subprocess, and is likely not a problem with pip.
error: legacy-install-failure_

**Comment by adamtuft** [comment]
In 71fefb9dcf I've replaced the dependency on `ipcqueue` with a different dependency `sysv_ipc`. This _should_ enable c-kernel to work on macOS, although I haven't been able to try it as I don't have access to a Mac. Incidentally, I have tested c-kernel on Windows, but only via the Windows Subsystem for Linux. I don't plan to provide native support for Windows itself.

To try this update:

```
git clone -b dev https://github.com/adamtuft/c-kernel.git
pip install ./c-kernel
python3 -m ckernel install AutoCompileKernel ckernel-dev "C-kernel (dev)" --cc gcc --cxx g++
```

Note that you will need to have `gcc` and `g++` installed to compile C/C++ respectively. On Mac I believe you can get these if you install Apple's developer tools. On Windows with WSL you should install your distro's developer package e.g. on Ubuntu, do `sudo apt update && sudo apt install build-essential`.

To check whether `gcc` was installed successfully, run `gcc --version`.

If you have a chance to test c-kernel on your machine, please let me know how you get on and what OS you used.

**Comment by mbukaea** [comment]
I can confirm that this installs on MacOS.I get an error when I try and run the kernel though. See info below

/opt/local/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/jupyter_server/transutils.py:13: FutureWarning: The alias `_()` will be deprecated. Use `_i18n()` instead.
  warnings.warn(warn_msg, FutureWarning)
[I 2023-09-22 09:56:45.631 LabApp] JupyterLab extension loaded from /opt/local/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/jupyterlab
[I 2023-09-22 09:56:45.631 LabApp] JupyterLab application directory is /opt/local/Library/Frameworks/Python.framework/Versions/3.11/share/jupyter/lab
[I 09:56:45.636 NotebookApp] The port 8888 is already in use, trying another port.
[I 09:56:45.637 NotebookApp] The port 8889 is already in use, trying another port.
[I 09:56:45.638 NotebookApp] Serving notebooks from local directory: /Users/wf5617
[I 09:56:45.638 NotebookApp] Jupyter Notebook 6.4.11 is running at:
[I 09:56:45.638 NotebookApp] http://localhost:8890/?token=2537711771055c550b75fa3ab930fe481667a6471ed1b6b5
[I 09:56:45.638 NotebookApp]  or http://127.0.0.1:8890/?token=2537711771055c550b75fa3ab930fe481667a6471ed1b6b5
[I 09:56:45.638 NotebookApp] Use Control-C to stop this server and shut down all kernels (twice to skip confirmation).
[C 09:56:45.660 NotebookApp] 
    
    To access the notebook, open this file in a browser:
        file:///Users/wf5617/Library/Jupyter/runtime/nbserver-62060-open.html
    Or copy and paste one of these URLs:
        http://localhost:8890/?token=2537711771055c550b75fa3ab930fe481667a6471ed1b6b5
     or http://127.0.0.1:8890/?token=2537711771055c550b75fa3ab930fe481667a6471ed1b6b5
0.00s - Debugger warning: It seems that frozen modules are being used, which may
0.00s - make the debugger miss breakpoints. Please pass -Xfrozen_modules=off
0.00s - to python to disable frozen modules.
0.00s - Note: Debugging will proceed. Set PYDEVD_DISABLE_FILE_VALIDATION=1 to disable this validation.
[I 09:56:59.523 NotebookApp] Creating new notebook in 
[I 09:57:01.345 NotebookApp] Kernel started: 3eaa6855-03f0-4859-9649-0847ee41d1d5, name: ckernel-dev
/opt/local/Library/Frameworks/Python.framework/Versions/3.11/bin/python3.11: No module named ckernel
[I 09:57:04.332 NotebookApp] KernelRestarter: restarting kernel (1/5), new random ports
/opt/local/Library/Frameworks/Python.framework/Versions/3.11/bin/python3.11: No module named ckernel
[I 09:57:07.352 NotebookApp] KernelRestarter: restarting kernel (2/5), new random ports
/opt/local/Library/Frameworks/Python.framework/Versions/3.11/bin/python3.11: No module named ckernel
[I 09:57:10.375 NotebookApp] KernelRestarter: restarting kernel (3/5), new random ports
/opt/local/Library/Frameworks/Python.framework/Versions/3.11/bin/python3.11: No module named ckernel
[I 09:57:13.395 NotebookApp] KernelRestarter: restarting kernel (4/5), new random ports
/opt/local/Library/Frameworks/Python.framework/Versions/3.11/bin/python3.11: No module named ckernel
[I 09:57:16.416 NotebookApp] KernelRestarter: restarting kernel (5/5), new random ports
/opt/local/Library/Frameworks/Python.framework/Versions/3.11/bin/python3.11: No module named ckernel
[W 09:57:19.436 NotebookApp] KernelRestarter: restart failed
[W 09:57:19.436 NotebookApp] Kernel 3eaa6855-03f0-4859-9649-0847ee41d1d5 died, removing from map.
[W 09:58:01.358 NotebookApp] Timeout waiting for kernel_info reply from 3eaa6855-03f0-4859-9649-0847ee41d1d5
[E 09:58:01.361 NotebookApp] Error opening stream: HTTP 404: Not Found (Kernel does not exist: 3eaa6855-03f0-4859-9649-0847ee41d1d5)

**Comment by adamtuft** [comment]
Ok, thanks for the error message! Can you please share the results of:

`python3 -m ckernel --version`

to check that ckernel is available to python, and also:

`jupyter kernelspec list --json`

Are you running the Jupyter server from within a virtual environment, or using Conda or something similar?

**Comment by mbukaea** [comment]
I just realised that I have multiple python 3 versions on my machine. I accidently installed it to one version and then tried to use it in a Jupyter notebook which used another version. I can confirm that the new kernel does work in MacOS.

**Comment by mbukaea** [comment]
Will this new kernel also allow me to create a clang version like the previous version?

**Comment by adamtuft** [comment]
Yes it will, just pass the `--cc` and `--cxx` options with your preferred compilers e.g.

```
python3 -m ckernel install AutoCompileKernel ckernel-clang "C/C++ (clang)" --cc clang --cxx clang++
```

**Comment by mbukaea** [comment]
Thanks for creating this. I've been trying and failing all week to get the xeus-cling kernel installed on my machine, so I could have a c++ kernel for my Jupyter notebook. Now that I have access to a c++ kernel through your c-kernel I plan to make use of it quite a lot.

**Comment by adamtuft** [comment]
This issue is fixed with #2

**Comment by adamtuft** [comment]
> Thanks for creating this. I've been trying and failing all week to get the xeus-cling kernel installed on my machine, so I could have a c++ kernel for my Jupyter notebook. Now that I have access to a c++ kernel through your c-kernel I plan to make use of it quite a lot.

You're very welcome, I'm really glad to hear that you like the kernel! Let me know how you get on with it!
